### ISSUE TITLE
Setting correct timezone when setting time

### ISSUE THREAD

**Issue by izivkov** [izivkov/GShockAPI#2]
We have now incorporated setting the Casio timezone and rules when setting time, in latest version of the API tagged 1.2.8. Here are some notes on the implementation.

When using the API, the app doesn't need to keep a list of world cities to set the time on the watch. It can simply provide the current timezone and the API will set the timezone and any applicable Daylight Savings Time rules on the watch. This way, when a user travels from one city to another and adjusts the time accordingly, the timezone and DST rules will be set accordingly.

2. In order to set timezone, we must provide data for register 0x1E, something like this:

      0x1E 00 **02 76 10 00 00**

From the [Gadgetbridge](https://codeberg.org/Freeyourgadget/Gadgetbridge/) project, we have information on what these values mean:
```
                   A  B   OFF DSTOFF DSTRULES
LOS ANGELES        A1 00  E0  04     01
DENVER             54 00  E4  04     01
CHICAGO            42 00  E8  04     01
NEW YORK           CA 00  EC  04     01
...
```

5. The [CasioTimeZone](https://github.com/izivkov/GShockAPI/blob/c0ec92f53efbf08f8046f088fe43cbc2fe42159e/api/src/main/java/org/avmedia/gshockapi/casio/CasioTimeZoneHelper.kt#L83C11-L83C24) class handles the code for setting the values of `time-zone`, `offset`, and `DST offset` on the watch. It takes a `time-zone` string in the form of "Pacific/Pago_Pago" and a Casio-specific DST rules code, and calculates the corresponding `offset` and `DST offset`. A table is then created of all possible `CasioTimeZone`'s that can be set on the watch.
7. The [findTimeZone](https://github.com/izivkov/GShockAPI/blob/c0ec92f53efbf08f8046f088fe43cbc2fe42159e/api/src/main/java/org/avmedia/gshockapi/casio/CasioTimeZoneHelper.kt#L167C9-L167C21) function will take a standard timezone obtained from the Android `Locale` and try to find the Casio-equivalent from the table. If not found, we create a synthetic `CasioTimeZone` from the `Offset` and `DST Offset` of the passed timezone, and a value of 0 for the Casio Rules. Most of the timezones not in the Casio table have no DST, so this is a reasonable assumption.
8. Since are are using [ZoneId](https://docs.oracle.com/javase/8/docs/api/java/time/ZoneId.html) API to obtain the TZ offset and DST offset, we don't have to worry about changes in the timezones, since presumably the API will change accordingly.
9. For setting DST value ON/OFF/AUTO, we update the `0x1D` register. The data looks something like this:

0x1D 00 **01 00 03 0C 01 0C 01 FF FF FF FF FF FF**

10. From Gadgetbridge, we see:
```
There are six clocks on the Casio GW-B5600
0 is the main clock
1-5 are the world clocks

0x1d 00 01 DST0 DST1 TZ0A TZ0B TZ1A TZ1B ff ff ff ff ff
0x1d 02 03 DST2 DST3 TZ2A TZ2B TZ3A TZ3B ff ff ff ff ff
0x1d 04 05 DST4 DST5 TZ4A TZ4B TZ5A TZ5B ff ff ff ff ff
DST: bitwise flags; bit0: DST on, bit1: DST auto
```
Here again, are are only concerned with the main clock, so we need to update the value of DST0. 
If the timezone has DST, we set this flag to ON | AUTO, or 3
If the timezone has no DST, we set the flag to 0

**Comment by johannesk** [comment]
Some comments:
>     5. The [CasioTimeZone](https://github.com/izivkov/GShockAPI/blob/c0ec92f53efbf08f8046f088fe43cbc2fe42159e/api/src/main/java/org/avmedia/gshockapi/casio/CasioTimeZoneHelper.kt#L83C11-L83C24) class handles the code for setting the values of time-zone, offset, and DST offset on the watch. It takes a time-zone string in the form of "Pacific/Pago_Pago" and a Casio-specific DST rules code, and calculates the corresponding offset and DST offset. A table is then created of all possible CasioTimeZone's that can be set on the watch.
The table already contains an outdated entry: Asia/Tehran abandoned DST in [2022](https://en.wikipedia.org/wiki/Daylight_saving_time_in_Iran). With your implementation you will have to update the table whenever some goverment decides to change their DST rules.

https://github.com/izivkov/GShockAPI/blob/c0ec92f53efbf08f8046f088fe43cbc2fe42159e/api/src/main/java/org/avmedia/gshockapi/casio/CasioTimeZoneHelper.kt#L85
`getDaylightSavings` only returns the correct value during summer time and returns 0 during winter time https://docs.oracle.com/javase/8/docs/api/java/time/zone/ZoneRules.html#getDaylightSavings-java.time.Instant-

>     8. For setting DST value ON/OFF/AUTO, we update the `0x1D` register. The data looks something like this:
ON/OFF and AUTO are independent bits. Your implementation sets both or none of the bits:
https://github.com/izivkov/GShockAPI/blob/c0ec92f53efbf08f8046f088fe43cbc2fe42159e/api/src/main/java/org/avmedia/gshockapi/io/TimeIO.kt#L49
This will give wrong results in two different cases:
1. A dstRule was found for this timezone, but it's currently wintertime. The implementation will set ON_AND_AUTO, although correct would be OFF_AND_AUTO. This will result in the watch substracting one hour from the set time.
2. No dstRule was found for this timezone, but the timezone has a DST offset and is currently in summer time.

**Comment by izivkov** [comment]
Thanks for your comments. I hope I'm not taking too much of your time.

I misinterpreted how getDaylightSavings() works, and have now corrected it.

I also make the assumption that if there is no DSTOffset (assuming it is calculated correctly), then there should not be dstRules. Thus for TEHRAN, the java.time API returns no DST offset, so I set the rules to 0. This should give me the correct values of the rules.

About the DST ON/OFF/AUTO flags:
Are you saying AUTO can only be ON if there are dstRules? I guess that makes sense.
I think the logic should be:

If DST rules present, set AUTO bit on, off otherwise.
If DST in effect, set ON bit
If DST not in effect, reset ON bit.

Lastly, if DST offset for this timezone is ZERO as per the java.time API,  I set the dstOffset to zero, even though the table shows 4 for many entries, like DELHI for example.
